File: 82_Remove Duplicates from Sorted List II.java
// Time complexity: O(n)
// Space complexity: O(1)

/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public ListNode deleteDuplicates(ListNode head) {
        ListNode vHead = new ListNode(0, head);
        ListNode cur = vHead;

        while (cur.next != null && cur.next.next != null) {
            int value = cur.next.val;
            // case: duplicate
            if (cur.next.next.val == value) {
                while (cur.next != null && cur.next.val == value) {
                    cur.next = cur.next.next;
                }
            }
            // case: not duplicate
            else cur = cur.next;
        }
        return vHead.next;
    }
}

File: 82_Remove Duplicates from Sorted List II.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def deleteDuplicates(self, head: Optional[ListNode]) -> Optional[ListNode]:
        # Remove all nodes w/ duplicates -> 2-ptr
        vHead = cur = ListNode(0, head)

        # We only worry about the next duplicates when >= 2 next nodes
        while cur.next and cur.next.next:
            value = cur.next.val
            # case: next duplicate -> skip all next-dups
            if cur.next.next.val == value:
                while cur.next and cur.next.val == value:
                    cur.next = cur.next.next
            # case: next not duplicate -> attach & proceed
            else:
                cur = cur.next
        return vHead.next
